File: topic/concurrency/thread/Puzzle.c
#include <stdbool.h>
#include <stdio.h>
#include <pthread.h>

int gAnswer = 0;
bool gAnswerReady = false;

void *task0(void *args) {
    gAnswerReady = true;
    gAnswer = 42;
    return NULL;
}

void *task1(void *args) {
    if (gAnswerReady) {
        printf("The meaning of life is : %d\n", gAnswer);
    } else {
        printf("I don't know the answer.\n");
    }
    return NULL;
}

int main() {
    pthread_t t0, t1;
    pthread_create(&t0, NULL, task0, NULL);
    pthread_create(&t1, NULL, task1, NULL);
    pthread_join(t0, NULL);
    pthread_join(t1, NULL);
    return 0;
}

File: topic/concurrency/thread/Puzzle.java
public class Puzzle {
    static boolean answerReady = false;
    static int answer = 0;
    static Thread t0 = new Thread() {
        public void run() {
            answer = 42;
            answerReady = true;
        }
    };

    static Thread t1 = new Thread() {
        public void run() {
            if (answerReady) {
                System.out.println("The meaning of life is : " + answer);
            } else {
                System.out.println("I don't know the answer.");
            }
        }
    };

    public static void main(String[] args)
        throws InterruptedException
    {
        // This example is bad.
        // cannot get "The meaning of life is : 0" result.
        // understand this is possible. But I cannot get this reuslt.
        t0.start();
        t1.start();
        t1.join();
        t0.join();
    }
}
